drake_ros_systems/src/tf_broadcaster_system.cpp

```diff
@@ -0,0 +1,139 @@
+  for (const auto & trigger : publish_triggers) {
+    if (
+      (trigger != TriggerType::kForced) &&
+      (trigger != TriggerType::kPeriodic) &&
+      (trigger != TriggerType::kPerStep))
+    {
+      throw std::invalid_argument(
+              "Only kForced, kPeriodic, or kPerStep are supported");
+    }
+  }
+
+  // Declare a forced publish so that any time Publish(.) is called on this
+  // system (or a Diagram containing it), a message is emitted.
+  if (publish_triggers.find(TriggerType::kForced) != publish_triggers.end()) {
+    this->DeclareForcedPublishEvent(&TfBroadcasterSystem::PublishFrames);
+  }
+
+  if (publish_triggers.find(TriggerType::kPeriodic) != publish_triggers.end()) {
+    if (publish_period <= 0.0) {
+      throw std::invalid_argument("kPeriodic requires publish_period > 0");
+    }
+    const double offset = 0.0;
+    this->DeclarePeriodicPublishEvent(
+      publish_period, offset,
+      &TfBroadcasterSystem::PublishFrames);
+  } else if (publish_period > 0) {
+    // publish_period > 0 without drake::systems::TriggerType::kPeriodic has no meaning and is
+    // likely a mistake.
+    throw std::invalid_argument("publish_period > 0 requires kPeriodic");
+  }
+
+  if (publish_triggers.find(TriggerType::kPerStep) != publish_triggers.end()) {
+    DeclarePerStepEvent(
+      drake::systems::PublishEvent<double>(
+        [this](
+          const drake::systems::Context<double> & context,
+          const drake::systems::PublishEvent<double> &) {
+          PublishFrames(context);
+        }));
+  }
+  // ^^^ Mostly copied from LcmPublisherSystem ^^^
+}
+
+TfBroadcasterSystem::~TfBroadcasterSystem()
+{
+}
+
+const drake::systems::InputPort<double> &
+TfBroadcasterSystem::get_graph_query_port() const
+{
+  return get_input_port(impl_->graph_query_port_index);
+}
+
+drake::systems::EventStatus
+TfBroadcasterSystem::PublishFrames(
+  const drake::systems::Context<double> & context) const
+{
+  const drake::geometry::QueryObject<double> & query_object =
+    get_graph_query_port().Eval<drake::geometry::QueryObject<double>>(context);
+  const drake::geometry::SceneGraphInspector<double> & inspector = query_object.inspector();
+  // TODO(hidmic): publish frame transforms w.r.t. to their parent frame
+  //               instead of the world frame when an API is made available.
+  if (inspector.num_frames() > 1) {
+    std::vector<geometry_msgs::msg::TransformStamped> transforms;
+    transforms.reserve(inspector.num_frames() - 1);
+    geometry_msgs::msg::TransformStamped transform;
+    transform.header.stamp =
+      rclcpp::Time() + rclcpp::Duration::from_seconds(context.get_time());
+    transform.header.frame_id = inspector.GetName(inspector.world_frame_id());
+    for (const drake::geometry::FrameId & frame_id : inspector.all_frame_ids()) {
+      if (frame_id == inspector.world_frame_id()) {
+        continue;
+      }
+      transform.child_frame_id = inspector.GetName(frame_id);
```

Comment:
What does this produce for frame names of identically-named frames in multiple model instances?

Reply:
Good question. I did not test that particular case. Having said that, AFAIK the `SceneGraph` does not seem to have a concept of model. Source systems like the `MultibodyPlant` use prefixes for geometry and frame names e.g. `robot_0::leg_link`. 

Thus,

> identically-named frames in multiple model instances?

shouldn't be a problem.

Reply:
On a related note, I considered substituting `::` with `/` to better align with ROS conventions for frame names, but I figured it'd be putting unnecessary cognitive load on a dev trying to correlate frames in the `SceneGraph` and `/tf`.

Reply:
Multiple model instances with identically-named frames is a very common case in anzu, so we need to make sure this works reasonably. This is part of the reason the anzu MbP+SG->Rviz process uses the MbP names, where model instances are known. (Also note that only some of these items (MbP bodies, MbP frames, SG frames) are guaranteed to have unique names, let alone names at all)

> considered substituting :: with / to better align with ROS conventions for frame names

We've used something like this in anzu
```
std::string GetTfFrameName(const std::string& model_instance_name, const std::string& element_name) {
  const std::regex mbp_separator_regex("::");
  std::string raw_name = model_instance_name + "/" + element_name;
  return std::regex_replace(raw_name, mbp_separator_regex, "/");
}
```
to do the conversion, and I think having ROS-like frame names where `::` always becomes `/` is well worth it to be clearer to users from the ROS side of things.

Reply:
> This is part of the reason the anzu MbP+SG->Rviz process uses the MbP names, where model instances are known.

> Also note that only some of these items (MbP bodies, MbP frames, SG frames) are guaranteed to have unique names, let alone names at all

I see. Indeed there's [no guarantee](https://github.com/RobotLocomotion/drake/blob/e59b7fc18dbe80b827d07e4a3283a0c87eda7021/geometry/geometry_state.cc#L544-L546) `SceneGraph` frame names are unique... Only frame IDs are unique.

@IanTheEngineer this is a bit of a problem. We cannot

> convert SceneGraph QueryObjects to [...] TF messages

if frame names are not unique. Well, we _could_ generate frame names like "frame_ID" but it is not very practical. We need additional information e.g. by querying the `MultibodyPlant` as @calderpg-tri suggests. I'm curious though, how is the scene described by a `SceneGraph` supposed to be visualized and introspected if the only non-ambiguous identification metadata is a number? I understand frame names become less useful as you scale into the hundredths or thousands of frames, but there must be some semantic metadata somewhere to make sense out of the scene, no?

> I think having ROS-like frame names where :: always becomes / is well worth it to be clearer to users from the ROS side of things.

That's fair. We can do the conversion.

Reply:
See `planning/rviz_geometry_visualizer.{cc, h}` and `planning/shape_description.{cc, h}` in anzu if you haven't already, to see how we've done this. We fall back to something like `<model_instance>/unnamed_frame_id_<frame_id>` in the case of a frame that doesn't have a name, but that's usually only a few frames in a complex system.

Reply:
In my mind, this is as much a ROS/RViz issue as it is for the `drake-ros` interface. Even in RViz, without special namespace handling, you cannot have multiple instances of robots (having the same names in their frames) robot publishing to `/tf`, and therefore visualized in RViz. It's not immediately obvious to me how we should handle this to make both Drake & `tf` happy. 

I think querying MBP for more information could work for now. We should make note if there is a particular MBP API that would make this process easier, we can propose it to the Drake Developers. From the `tf`/RViz side, we could bring back to life something akin to `tf_prefix`, which was apparently only [deprecated due to a lack of development cycles](http://wiki.ros.org/tf2/Migration#tf_prefix_backwards_compatibility).

Reply:
Alright, https://github.com/RobotLocomotion/drake-ros/pull/2/commits/b365190165eee7d24751b273fe432b9d96609b22 and https://github.com/RobotLocomotion/drake-ros/pull/2/commits/1f69048dd13d918cbdaed04471272fba7d786f8d now allows `MultibodyPlant` instances in the scene to be registered with the `RvizVisualizer` to help refine frame IDs and marker namespaces.